cpp/ql/src/Critical/OverflowDestination.ql

```diff
@@ -45,9 +48,49 @@ predicate sourceSized(FunctionCall fc, Expr src) {
   )
 }
 
-from FunctionCall fc, Expr vuln, Expr taintSource
+predicate readsVariable(LoadInstruction load, Variable var) {
+  load.getSourceAddress().(VariableAddressInstruction).getASTVariable() = var
+}
+
+predicate hasUpperBoundsCheck(Variable var) {
+  exists(RelationalOperation oper, VariableAccess access |
+    oper.getAnOperand() = access and
+    access.getTarget() = var and
+    // Comparing to 0 is not an upper bound check
+    not oper.getAnOperand().getValue() = "0"
+  )
+}
+
+predicate nodeIsBarrierEqualityCandidate(DataFlow::Node node, Operand access, Variable checkedVar) {
+  readsVariable(node.asInstruction(), checkedVar) and
+  any(IRGuardCondition guard).ensuresEq(access, _, _, node.asInstruction().getBlock(), true)
+}
+
+class OverflowDestinationConfig extends TaintTracking::Configuration {
+  OverflowDestinationConfig() { this = "OverflowDestinationConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof FlowSource }
+
+  override predicate isSink(DataFlow::Node sink) { sourceSized(_, sink.asExpr()) }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    exists(Variable checkedVar |
+      readsVariable(node.asInstruction(), checkedVar) and
+      hasUpperBoundsCheck(checkedVar)
+    )
+    or
+    exists(Variable checkedVar, Operand access |
+      readsVariable(access.getDef(), checkedVar) and
+      nodeIsBarrierEqualityCandidate(node, access, checkedVar)
+    )
```

Comment:
This is a copy of `nodeIsBarrier` from `DefaultTaintTracking.qll`, right?  If we're emulating that, do `nodeIsBarrierIn` and `additionalTaintStep` from `DefaultTaintTracking.qll` matter as well?

Reply:
Also it might be good to have a note explaining their history and that they may or may not actually be helping the query at this point - so that future maintainers can potentially clean things up without being overly confused/paranoid.

Reply:
> This is a copy of `nodeIsBarrier` from `DefaultTaintTracking.qll`, right? If we're emulating that, do `nodeIsBarrierIn` and `additionalTaintStep` from `DefaultTaintTracking.qll` matter as well?

Yep, it's a copy of `nodeIsBarrier` from `DefaultTaintTracking.qll`. `isAdditionalTaintStep` shouldn't matter as those taint steps should already be provided by `semmle.code.cpp.ir.dataflow.internal.TaintTrackingUtil.qll`.

`nodeIsBarrierIn` is more complicated. I hope we don't need them as I don't like the whole concept of `unpredictable` instructions. Let's see what DCA tells us about the result changes (if any).

Reply:
> Also it might be good to have a note explaining their history and that they may or may not actually be helping the query at this point - so that future maintainers can potentially clean things up without being overly confused/paranoid.

It definitely is changing the query results. It's blocking a lot of flow. Most of them are flows we do want to block, but sometimes it's also the cause of `[NOT DETECTED]` results.